pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,318 @@
+package cachemanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/metrics"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil/aggregator"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/target"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("cache-manager")
+
+type Config struct {
+	Opa              syncutil.OpaDataClient
+	SyncMetricsCache *syncutil.MetricsCache
+	Tracker          *readiness.Tracker
+	ProcessExcluder  *process.Excluder
+	Registrar        *watch.Registrar
+	WatchedSet       *watch.Set
+	GVKAggregator    *aggregator.GVKAgreggator
+	Reader           client.Reader
+}
+
+type CacheManager struct {
+	// the processExcluder and the gvkAggregator define what the underlying
+	// cache should look like. we refer to those two as "the spec"
+	processExcluder *process.Excluder
+	gvkAggregator   *aggregator.GVKAgreggator
+	// mu guards access to any part of the spec above
+	mu sync.RWMutex
+
+	opa              syncutil.OpaDataClient
+	syncMetricsCache *syncutil.MetricsCache
+	tracker          *readiness.Tracker
+	registrar        *watch.Registrar
+	watchedSet       *watch.Set
+	replayTicker     time.Ticker
+	reader           client.Reader
+	excluderChanged  bool
+}
+
+func NewCacheManager(config *Config) (*CacheManager, error) {
+	if config.WatchedSet == nil {
+		return nil, fmt.Errorf("watchedSet must be non-nil")
+	}
+	if config.Registrar == nil {
+		return nil, fmt.Errorf("registrar must be non-nil")
+	}
+	if config.ProcessExcluder == nil {
+		return nil, fmt.Errorf("processExcluder must be non-nil")
+	}
+	if config.Tracker == nil {
+		return nil, fmt.Errorf("tracker must be non-nil")
+	}
+	if config.Reader == nil {
+		return nil, fmt.Errorf("reader must be non-nil")
+	}
+
+	cm := &CacheManager{
+		opa:              config.Opa,
+		syncMetricsCache: config.SyncMetricsCache,
+		tracker:          config.Tracker,
+		processExcluder:  config.ProcessExcluder,
+		registrar:        config.Registrar,
+		watchedSet:       config.WatchedSet,
+		reader:           config.Reader,
+	}
+
+	cm.gvkAggregator = aggregator.NewGVKAggregator()
+
+	cm.replayTicker = *time.NewTicker(3 * time.Second)
+
+	return cm, nil
+}
+
+func (c *CacheManager) Start(ctx context.Context) error {
+	go c.updateDatastore(ctx)
+
+	<-ctx.Done()
+	return nil
+}
+
+// AddSource adjusts the watched set of gvks according to the newGVKs passed in
+// for a given sourceKey.
+func (c *CacheManager) AddSource(ctx context.Context, sourceKey aggregator.Key, newGVKs []schema.GroupVersionKind) error {
```

Comment:
If we are wanting net-new GVKs to be watched ASAP, they should not be waiting on the ticker.